cli/command/image/build.go

```diff
@@ -235,6 +247,47 @@ func runBuild(dockerCli *command.DockerCli, options buildOptions) error {
 		}
 	}
 
+	// replace Dockerfile if added dynamically
+	if dockerfileCtx != nil {
+		file, err := ioutil.ReadAll(dockerfileCtx)
+		dockerfileCtx.Close()
+		if err != nil {
+			return err
+		}
+		now := time.Now()
+		hdrTmpl := &tar.Header{
+			Mode:       0600,
+			Uid:        0,
+			Gid:        0,
+			ModTime:    now,
+			Typeflag:   tar.TypeReg,
+			AccessTime: now,
+			ChangeTime: now,
+		}
+		randomName := ".dockerfile." + stringid.GenerateRandomID()[:12]
```

Comment:
I know its really really unlikely, but to avoid a flaky situation, can you just check to make sure there isn't a file with this name already in the build context? And if so, try again.